src/operator/tensor/elemwise_unary_op_trig.cc

```diff
@@ -139,7 +139,31 @@ The storage type of ``tan`` output depends upon the input storage type:
 )code" ADD_FILELINE)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseOut{ "_backward_tan" });
 
-MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_tan, unary_bwd<mshadow_op::tan_grad>);
+MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_tan, unary_bwd<mshadow_op::tan_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+  [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+      // NodeEntry{n} : y_grad * f'(x)
+      // n->inputs[0] : y_grad
+      // n->inputs[1] : f(x) = tan(x)
+      // ograds[0] : head_grads
+      // f'(x) = sec^2(x)
+      // f''(x) = 2 * f'(x) * f(x)
+      const std::unordered_map<std::string, std::string> args = {{"scalar", "2.0"}};
+      auto two_y = MakeNode("_mul_scalar", n->attrs.name + "_mul_two", {n->inputs[1]}, &args, &n);
+      auto grad_grad_mid = MakeNode("elemwise_mul", n->attrs.name + "_grad_mul",
```

Comment:
Can we clarify / add a comment on why is correct to multiply by y_grad (the first head gradient?) again? This would help readers as is not obvious, as well as the very non-obvious implicit multiplication by f'(x) it compounds.

Reply:
Makes sense. Thanks. Will get to it.

Reply:
I have updated the comment. See if it is okay? Or maybe the phrasing can be improved.
Thanks

Reply:
Thanks. About the outputs, I think we should write some documentation explaining what we are doing as I find it non trivial. Can you help me understand the y_grad_grad (first output)?

If you want, we can move the conversation to the dev list or slack, as the PR LGTM.

![IMG_20190711_122948__01](https://user-images.githubusercontent.com/928489/61079539-f1b55600-a3d7-11e9-9bfe-36b45706e47e.jpg)

Reply:
Ok I clarified with @apeforest , this makes sense now.